great_expectations/expectations/core/expect_query_results_to_match_source.py

```diff
@@ -192,5 +207,70 @@ def _validate(
                 "result": {
                     "unexpected_count": unexpected_count,
                     "unexpected_percent": unexpected_percent,
+                    "details": {
+                        "missing_rows": missing_rows,
+                        "unexpected_rows": unexpected_rows,
+                    },
                 },
             }
+
+    @classmethod
+    def _compute_row_data(
+        cls,
+        col_names: list[str],
+        frequency_map: Counter[tuple],
+    ) -> list[dict[str, Any]]:
+        """Given a frequency map of values and a list of their keys, compute a list of
+        column-name -> column value dictionaries.
+
+        Example:
+          - col_names: ["a", "b"]
+          - frequency_map: Counter({
+                (1, 2): 2,
+                (3, 4): 1,
+            })
+
+          -> [
+                {"a": 1, "b": 2},
+                {"a": 1, "b": 2},
+                {"a": 3, "b": 4},
+             ]
+        """
+        row_values = sorted(frequency_map.elements(), key=cmp_to_key(cls._null_safe_tuple_compare))
+
+        return [
+            {col_names[i]: row_values[i] for i in range(len(col_names))}
+            for row_values in row_values
+        ]
+
+    @classmethod
+    def _get_column_names_from_result(
+        cls,
+        results_list: list[dict[str, Any]],
+    ) -> list[str]:
+        """Get the list of columns from a result list.
+
+        NOTE: Order matters here, and we rely on python 3.7's deterministic ordering of results.
+        """
+        if results_list:
+            return list(results_list[0].keys())
+        else:
+            return []
+
+    @classmethod
+    def _null_safe_tuple_compare(
```

Comment:
Maybe documentation on line 239 will make it clearer how to interpret this method. "compare" feels a bit too general.

Reply:
I think "compare" is the right term here, but LMK if you have another suggestion and I can rename.